src/test/fio/fio_ceph_objectstore.cc

```diff
@@ -0,0 +1,396 @@
+  cct = boost::intrusive_ptr<CephContext>(g_ceph_context, false);
+
+  // create the ObjectStore
+  os.reset(ObjectStore::create(g_ceph_context,
+                               g_conf->osd_objectstore,
+                               g_conf->osd_data,
+                               g_conf->osd_journal));
+  if (!os)
+    throw std::runtime_error("bad objectstore type " + g_conf->osd_objectstore);
+
+  int r = os->mkfs();
+  if (r < 0)
+    throw std::system_error(-r, std::system_category(), "mkfs failed");
+
+  r = os->mount();
+  if (r < 0)
+    throw std::system_error(-r, std::system_category(), "mount failed");
+}
+
+Engine::~Engine()
+{
+  os->umount();
+}
+
+
+struct Collection {
+  spg_t pg;
+  coll_t cid;
+  ObjectStore::Sequencer sequencer;
+
+  // use big pool ids to avoid clashing with existing collections
+  static constexpr int64_t MIN_POOL_ID = 0x0000ffffffffffff;
+
+  Collection(const spg_t& pg)
+    : pg(pg), cid(pg), sequencer(stringify(pg)) {}
+};
+
+struct Object {
+  ghobject_t oid;
+  Collection& coll;
+
+  Object(const char* name, Collection& coll)
+    : oid(hobject_t(name, "", CEPH_NOSNAP, coll.pg.ps(), coll.pg.pool(), "")),
+      coll(coll) {}
+};
+
+/// treat each fio job like a separate pool with its own collections and objects
+struct Job {
+  Engine* engine; //< shared ptr to the global Engine
+  std::vector<Collection> collections; //< spread objects over collections
+  std::vector<Object> objects; //< associate an object with each fio_file
+  std::vector<io_u*> events; //< completions for fio_ceph_os_event()
+  const bool unlink; //< unlink objects on destruction
+
+  Job(Engine* engine, const thread_data* td);
+  ~Job();
+};
+
+Job::Job(Engine* engine, const thread_data* td)
+  : engine(engine),
+    events(td->o.iodepth),
+    unlink(td->o.unlink)
+{
+  // use the fio thread_number for our unique pool id
+  const uint64_t pool = Collection::MIN_POOL_ID + td->thread_number;
+
+  // create a collection for each object, up to osd_pool_default_pg_num
+  uint32_t count = g_conf->osd_pool_default_pg_num;
+  if (count > td->o.nr_files)
+    count = td->o.nr_files;
+
+  assert(count > 0);
+  collections.reserve(count);
+
+  const int split_bits = cbits(count - 1);
+
+  ObjectStore::Transaction t;
+  for (uint32_t i = 0; i < count; i++) {
+    auto pg = spg_t{pg_t{i, pool}};
+    collections.emplace_back(pg);
+
+    auto& cid = collections.back().cid;
+    if (!engine->os->collection_exists(cid))
+      t.create_collection(cid, split_bits);
+  }
+
+  const uint64_t file_size = td->o.size / max(1u, td->o.nr_files);
+
+  // create an object for each file in the job
+  for (uint32_t i = 0; i < td->o.nr_files; i++) {
+    auto f = td->files[i];
+    f->real_file_size = file_size;
+    f->engine_data = i;
+
+    // associate each object with a collection in a round-robin fashion
+    auto& coll = collections[i % collections.size()];
+    objects.emplace_back(f->file_name, coll);
```

Comment:
@cbodley are you sure that it's touch/truncate that causes the iops drop? It's not evident from @tanghaodong25 comment IMHO. I suppose it's object size that might impact that. In case of large object one has pretty large metadata per object and each write op on it triggers metadata serialization and db update.